### PR TITLE
Support Encryption

### LINKED ISSUE #85: Support for encryption for storage class
**Is your feature request related to a problem?/Why is this needed**

As a user concerned by data storage security, I would like to be able to encrypt my data (either through block or mount) to prevent some case of unauthorized read from BSU volumes.

/feature

**Describe the solution you'd like in detail**

Be able to provide a secret at storage class declaration similarly to what [portworx](https://docs.portworx.com/key-management/kubernetes-secrets/pvc-encryption-using-csi/) has done.

Secret can be passed to CSI thanks to [secrets requirements](https://github.com/container-storage-interface/spec/blob/master/spec.md#secrets-requirements).

This issue focus on encryption at storage class level, a future enhancement would be to allow specific secret definition at PersistentVolumeClaim level.

### PR DESCRIPTION
**Pull request by outscale-mdr**
This PR adds the feature to support encryption of disk using LUKS.

Close #85

### CODE REVIEW

File: pkg/driver/node.go [hunk 1]
```diff
@@ -91,8 +91,7 @@ func newNodeService() nodeService {
 }
 
 func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
-	klog.V(4).Infof("NodeStageVolume: called with args %+v", *req)
-	klog.V(4).Infof("%#v", req.GetPublishContext())
+	klog.V(4).Infof("NodeStageVolume: called with args  {VolumeId:%+v, PublishContext:%+v, StagingTargetPath:%+v, VolumeCapability:%+v, VolumeContext:%+v}", req.VolumeId, req.PublishContext, req.StagingTargetPath, req.VolumeCapability, req.VolumeContext)
```
Comment (jerome-jutteau) [inline pkg/driver/node.go:94]: `VolumeContext` may contains encryption key?

File: docs/README.md [hunk 2]
```diff
@@ -30,13 +30,18 @@ The following CSI gRPC calls are **not yet** implemented:
...
+| "csi.storage.k8s.io/node-stage-secret-name"      | string                |         | The name of the secret  (See [template](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#node-stage-secret))     |
+| "csi.storage.k8s.io/node-stage-secret-namespace" | string                |         | The namespace of the secret (See [template](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#node-stage-secret)) |
+| "kmsKeyId"                                       | string                |         | Not yet supported                                                                                                                                |
+| "luks-cipher"                                    | string                |         | LUKS encryption cipher to use  (See [doc](https://wiki.archlinux.org/title/dm-crypt/Device_encryption#Encryption_options_for_LUKS_mode))                    |
+| "luks-hash"                                      | string                |         | Derivation Password hash algorithm (See [doc](https://wiki.archlinux.org/title/dm-crypt/Device_encryption#Encryption_options_for_LUKS_mode))     |
```
Comment (jerome-jutteau) [inline docs/README.md:43]: Maybe mention that the default hash and cypher depends of installed luks version.
Also, maybe point to official documentation if possible (https://gitlab.com/cryptsetup/cryptsetup/blob/master/README.md)

**Review by jerome-jutteau — CHANGES_REQUESTED** [review]

File: pkg/driver/luks_util.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+func LuksOpen(exec Mounter, devicePath string, encryptedDeviceName string, passphrase string) (bool, error) {
+	if ok, err := exec.ExistsPath("/dev/mapper/" + encryptedDeviceName); err == nil && ok {
+		klog.V(4).Info("luks volume is already open")
+		return true, nil
+	}
+	passphraseCmd := exec.Command("cryptsetup", "-v", "--type=luks2", "--batch-mode", "luksOpen", devicePath, encryptedDeviceName)
+	passwordReader := strings.NewReader(passphrase)
+	passphraseCmd.SetStdin(passwordReader)
+	if out, err := passphraseCmd.CombinedOutput(); err != nil {
+		return false, fmt.Errorf("err: %v, output: %v", err, out)
```
[the remark this answers is not in the document]
Comment (outscale-mdr) [reply pkg/driver/luks_util.go:64]: Cryptsetup will not output the secret even if it is a wrong one

File: pkg/driver/node.go [hunk 4]
```diff
@@ -180,17 +182,78 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
...
+		encryptedDevicePath = fmt.Sprintf("/dev/mapper/%v", encryptedDeviceName)
+
+		if device == encryptedDevicePath {
+			klog.V(4).Infof("NodeStageVolume: volume=%q already staged (encryption)", volumeID)
+			return &csi.NodeStageVolumeResponse{}, nil
+		}
+
+		passphrase, ok := req.Secrets[LuksPassphraseKey]
+		if !ok {
+			return nil, status.Error(codes.InvalidArgument, "No Secrets have been provided")
```
[the remark this answers is not in the document]
Comment (outscale-mdr) [reply pkg/driver/node.go:205]: It is better, yes

File: pkg/driver/node.go [hunk 5]
```diff
@@ -180,17 +182,78 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
...
+		}
+
+		// Check passphrase
+		if ok := d.mounter.CheckLuksPassphrase(source, passphrase); !ok {
+			msg := fmt.Sprintf("error while checking passphrase to %v, err: %v", volumeID, err)
+			return nil, status.Error(codes.Internal, msg)
+		}
+
+		// Open disk
+		// TODO Check it has already been opened
```
[the remark this answers is not in the document]
Comment (outscale-mdr) [reply pkg/driver/node.go:229]: old one, removing it

File: pkg/driver/node.go (hunk 1 above)
Comment (outscale-mdr) [reply pkg/driver/node.go:94]: VolumeContext will only contain configuration luks, the secret is provided in `NodeStage` step

**Review by jerome-jutteau — APPROVED** [review]

**Pull request merged**